File: apps/android/feature/presence/src/main/kotlin/com/fish/android/feature/presence/PresenceIndicator.kt
package com.fish.android.feature.presence

import androidx.compose.material3.Icon
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.semantics.clearAndSetSemantics
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.foundation.layout.size
import com.fish.android.core.designsystem.FishIcons
import com.fish.android.core.designsystem.FishTheme
import com.fish.android.data.presence.PresenceDisplayStatus

@Composable
fun PresenceIndicator(
    status: PresenceDisplayStatus,
    label: String,
    modifier: Modifier = Modifier,
    decorative: Boolean = false,
) {
    val (icon, color) = status.visuals()
    val semantics = if (decorative) {
        Modifier.clearAndSetSemantics { }
    } else {
        Modifier.semantics { contentDescription = label }
    }
    Icon(
        imageVector = icon,
        contentDescription = null,
        tint = color,
        modifier = modifier
            .then(semantics)
            .size(FishTheme.sizes.presenceIndicatorSmall),
    )
}

@Composable
private fun PresenceDisplayStatus.visuals(): Pair<ImageVector, Color> = when (this) {
    PresenceDisplayStatus.Online -> FishIcons.CircleFilled to FishTheme.colors.presenceOnline
    PresenceDisplayStatus.Idle -> FishIcons.Moon to FishTheme.colors.presenceIdle
    PresenceDisplayStatus.Away -> FishIcons.Clock to FishTheme.colors.presenceAway
    PresenceDisplayStatus.Busy -> FishIcons.CircleMinus to FishTheme.colors.presenceBusy
    PresenceDisplayStatus.Invisible -> FishIcons.EyeOff to FishTheme.colors.presenceOffline
    PresenceDisplayStatus.Offline -> FishIcons.OutlineCircle to FishTheme.colors.presenceOffline
}

File: apps/ios/FishKit/Sources/UIComponents/Identity/PresenceIndicator.swift
import DesignSystem
import SwiftUI

/// The shared display vocabulary for presence. Subjects never show
/// `invisible` — the server sanitizes them to offline; only the owner's own
/// status can display it.
public enum PresenceDisplayStatus: Sendable, Equatable, CaseIterable {
    case online
    case idle
    case away
    case busy
    case invisible
    case offline
}

/// Status glyph — always shape plus color, never color alone, mirroring the
/// web indicator icon-for-icon. Decorative by default because adjacent text
/// usually carries the status label.
public struct PresenceIndicator: View {
    private let status: PresenceDisplayStatus
    private let label: String
    private let isDecorative: Bool

    public init(
        status: PresenceDisplayStatus,
        label: String,
        isDecorative: Bool = true
    ) {
        self.status = status
        self.label = label
        self.isDecorative = isDecorative
    }

    public var body: some View {
        status.icon.image
            .frame(
                width: Metrics.presenceIndicatorSmall,
                height: Metrics.presenceIndicatorSmall
            )
            .foregroundStyle(status.color)
            .accessibilityHidden(isDecorative)
            .accessibilityLabel(isDecorative ? "" : label)
    }
}

extension PresenceDisplayStatus {
    var icon: Icon {
        switch self {
        case .online: .circleFilled
        case .idle: .moonFilled
        case .away: .clock
        case .busy: .circleMinus
        case .invisible: .eyeOff
        case .offline: .circle
        }
    }

    var color: Color {
        switch self {
        case .online: Palette.presenceOnline
        case .idle: Palette.presenceIdle
        case .away: Palette.presenceAway
        case .busy: Palette.presenceBusy
        case .invisible: Palette.presenceOffline
        case .offline: Palette.presenceOffline
        }
    }
}
